[#993] make web app url more flexible

zhiyuli opened the pull request:
This PR tries to address #993. We introduced some pre-defined terms (more terms can be added as needed):
${HS_RES_ID}  --> resource id
${HS_RES_TYPE}  --> resource type
${HS_USR_NAME}  --> username

User can use these pre-defined terms to construct a url template, like:
http://www.myapps.com/app1/?res_id=${HS_RES_ID}&app_ver=2.0&res_type=${HS_RES_TYPE}
then our code replaces these terms with real values. This allows user to create more flexible urls that fit their apps.

hydrobuild commented:
Can one of the admins verify this patch?

pkdash commented:
@selimnairb Can you review this?

danames commented:
Drew. I think the suggestion used curly braces, not parentheses. Does that
matter?

On Mon, Feb 22, 2016, 9:03 PM Pabitra Dash <notifications@github.com> wrote:

> @selimnairb <https://github.com/selimnairb> Can you review this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hydroshare/hydroshare/pull/1003#issuecomment-187515947>
> .
>

zhiyuli commented:
@danames  I think it is ok. I didn't use regex, just used python string.replace().  As curly braces is a special character in python string, I simply changed it to parentheses.

danames commented:
@yanliu-chn <https://github.com/yanliu-chn> does this solution work for you
- particularly using parentheses instead of curly braces - since curly
braces make the string replace more complicated.

On Tue, Feb 23, 2016 at 8:18 AM Drew Li <notifications@github.com> wrote:

> @danames <https://github.com/danames> I think it is ok. I didn't use
> regex, just used python string.replace(). As curly braces is a special
> character in python string, I simply changed it to parentheses.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hydroshare/hydroshare/pull/1003#issuecomment-187737481>
> .
>

yanliu-chn commented:
Hi @zhiyuli and @danames ,

Either is fine, I think. However, I don't understand why the replace() function cares about curly brackets. You might have used some formatting function that interprets { as a python code snippet in other places.

But, when an app contributor creates and saves an app resource, you probably still need regex to validate the list of HS parameters defined in the url in order to catch misspelled parameter names, invalid parameters (not allowed but exists, due to access control, e.g.).

zhiyuli commented:
@yanliu-chn @danames 
yes, I tried to separate the whole thing into two functions:
1) term definitions
2) replace terms with real values
so the (2) needs to use (1) to create search patterns dynamically. 
The code is pretty short: https://github.com/hydroshare/hydroshare/blob/993-make-webapp-url-more-flexible/hs_tools_resource/utils.py

For the validation part, I just use django built-in type forms.URLField() to check it. We could add more sophisticated validations later...

Thanks
Drew

pkdash commented:
@selimnairb Can you review this?

pkdash commented:
@selimnairb Please review.

selimnairb commented on hs_tools_resource/utils.py:
```diff
@@ -0,0 +1,24 @@
+# $(HS_RES_TYPE) resource id
+# $(HS_USR_NAME) user name
+
+def HS_term_dict(res_obj, usr_obj):
```
Please document the purpose of this function.

zhiyuli commented:
@selimnairb  I am making some changes to this branch right now. I will push the latest code in about one hour. thanks

selimnairb commented on hs_tools_resource/utils.py:
```diff
@@ -0,0 +1,24 @@
+
+
+def parse_app_url(url_string, term_dict):
+
```
Please document the purpose of this function as well as input parameters and return data.

selimnairb commented on hs_tools_resource/utils.py:
```diff
@@ -0,0 +1,24 @@
+
+    new_url_string = url_string
+    for key, value in term_dict.iteritems():
+        new_url_string = new_url_string.replace("$({0})".format(key), str(value))
```
Why not use [string.Template](https://docs.python.org/2.7/library/string.html#string.Template), which has a (substitute)[https://docs.python.org/2.7/library/string.html#string.Template.substitute] method that does the substitution against a dict?

zhiyuli replied:
@selimnairb Thanks for telling me this module. It is handy!

selimnairb commented on hs_tools_resource/utils.py:
```diff
@@ -0,0 +1,24 @@
+# $(HS_RES_ID)  resource id
```
These should be removed.

selimnairb commented:
I noted some minor changes I'd like to see, but otherwise this looks good.

zhiyuli commented:
@selimnairb 
We added a function "get_hs_term_dict()" in BaseResource, which defines some common terms like HS_RES_ID-->res _id and HS_RES_TYPE-->res type. Other resource types can override this function to add more resource-specific terms.

Also, I addressed your review comments: added documents for major functions, and changed to use   string.Template module to parse the url template.

Will you please do a final review and give a +1 if it is good.
Thanks

selimnairb commented on hs_tools_resource/utils.py:
```diff
@@ -0,0 +1,18 @@
+
+    new_url_string = url_template_string
+    for term_dict in term_dict_list:
+        new_url_string = Template(new_url_string).safe_substitute(term_dict)
```
It would be better to combine the dicts into one, then call substitute(), which will throw an exception if there are templates that are not matched.  Otherwise, you will blindly return an invalid URL (i.e. '{' and '}' can't appear in a URL unless they are escaped).  We shouldn't allow users to generate such URLs.

selimnairb commented:
@zhiyuli I made one more comment.  Once this is fixed +1.

ironfroggy commented on hs_core/models.py:
```diff
@@ -1362,7 +1362,19 @@ def can_have_multiple_files(cls):
 
+        # NOTES FOR ANY SUBCLASS OF THIS CLASS TO OVERRIDE THIS FUNCTION:
+        # resource types that inherit this class should add/merge their resource-specific HS Terms
+        # into this dict
```
Please use a doctstring for this documentation

ironfroggy commented on hs_geographic_feature_resource/models.py:
```diff
@@ -89,6 +89,14 @@ def can_have_multiple_files(cls):
         # can have more than one files
         return True
 
+    # demo how to add resource-specific HS terms
```
why is this called a "demo"?

ironfroggy commented on hs_geographic_feature_resource/models.py:
```diff
@@ -89,6 +89,14 @@ def can_have_multiple_files(cls):
+        # get existing hs_term_dict from base class
+        hs_term_dict = super(GeographicFeatureResource, self).get_hs_term_dict()
+        geometryinformation = self.metadata.geometryinformation.all().first()
+        hs_term_dict["HS_GFR_FEATURE_COUNT"] = geometryinformation.featureCount if geometryinformation is not None else 0
```
this line is long and hard to read, please break it up into a normal if statement

ironfroggy commented on hs_tools_resource/forms.py:
```diff
@@ -33,7 +33,7 @@ class Meta:
 
 class UrlBaseValidationForm(forms.Form):
-    value = forms.CharField(max_length=1024)
+    value = forms.URLField(max_length=1024)
```
Did you try to run `makemigrations` after this change?

zhiyuli commented:
@selimnairb your comment has been addressed. thanks

@ironfroggy  Thanks for double checking this PR.  I just checked that the change to the forms.py didn't create a new migration file. Your other comments have been taken care of as well.  thanks

selimnairb commented on hs_core/page_processors.py:
```diff
@@ -68,19 +69,16 @@ def get_page_context(page, user, resource_edit=False, extended_metadata_layout=N
+                        hs_term_dict_user = {}
+                        hs_term_dict_user["HS_USR_NAME"] = request.user.username if request.user.is_authenticated() else "anonymous"
+                        tool_url_new = parse_app_url_template(tool_url, [content_model.get_hs_term_dict(), hs_term_dict_user])
+                        if tool_url_new is not None:
```
What if tool_url_new is None, what do we do then?

zhiyuli replied:
@selimnairb  currently we just not display that webapp on other resource landing pages

selimnairb replied:
Okay.

selimnairb commented on hs_tools_resource/utils.py:
```diff
@@ -0,0 +1,22 @@
+
+        new_url_string = Template(new_url_string).substitute(merged_term_dic)
+    except Exception as ex:
+        new_url_string = None
```
This exception should be logged:

At root of the module:

```
import logging

logger = logging.getLogger(__name__)
```
In your function:

logger.error('DESCRIBE THE ERROR')

zhiyuli replied:
@selimnairb 
is "__name__" a right parameter for logging.getLogger(). where should I find the log it outputs.

zhiyuli replied:
@selimnairb I found it. it got logged in hydroshare.log.

zhiyuli commented:
@selimnairb 
I just added a logger. thanks

selimnairb commented:
+1 once others' comments have been resolved.

zhiyuli commented:
@selimnairb Thanks for reviewing this.

@ironfroggy Your comments have been addressed as well. Will you give a +1? thanks

ironfroggy commented on hs_tools_resource/utils.py:
```diff
@@ -0,0 +1,26 @@
+
+logger = logging.getLogger(__name__)
+
+def parse_app_url_template(url_template_string, term_dict_list=[]):
```
do not use mutable values as defaults in function parameters, as they are shared between *all* calls to the function. Use a tuple instead.

zhiyuli commented:
@mjstealey this pr is ready to go. thanks

mjstealey commented:
> this pr is ready to go. thanks

@zhiyuli - you've checked [this](https://github.com/hydroshare/hydroshare/pull/1003/files#r54935751)?

zhiyuli commented:
@mjstealey   We already got a +1 from @selimnairb above,  and another +1 from @ironfroggy in hipchat.

zhiyuli commented:
@mjstealey oh i just realized what you are asking. Yes, I checked that change, it did not create a new migration file. thanks

mjstealey commented:
> it did not create a new migration file.

Yep, that is what I was looking for
